types/express-serve-static-core/index.d.ts

```diff
@@ -96,16 +98,23 @@ export type RequestHandlerParams<
     | ErrorRequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>
     | Array<RequestHandler<P> | ErrorRequestHandler<P>>;
 
-export type RouteParameterNames<Route extends string> =
-    string extends Route
-        ? string
-        : Route extends `${string}:${infer Param}/${infer Rest}`
-        ? (Param | RouteParameterNames<Rest>)
-        : (Route extends `${string}:${infer LastParam}` ? LastParam : never);
-
-export type RouteParameters<T extends string> = {
-    [key in RouteParameterNames<T>]: string
-};
+type GetRouteParameter<RouteAfterColon extends string> = RouteAfterColon extends `${infer Char}${infer Rest}`
+    ? Char extends '/' | '-' | '.'
+        ? ''
+        : `${Char}${GetRouteParameter<Rest>}`
+    : RouteAfterColon;
+
+// prettier-ignore
+export type RouteParameters<Route extends string> = string extends Route
+    ? ParamsDictionary
+    : Route extends `${string}(${string}`
+        ? ParamsDictionary //TODO: handling for regex parameters
+        : Route extends `${string}:${infer Rest}`
+            ? (GetRouteParameter<Rest> extends `${infer ParamName}?`
+                ? { [P in ParamName]?: string }
+                : { [P in GetRouteParameter<Rest>]: string }) &
+                (Rest extends `${GetRouteParameter<Rest>}${infer Next}` ? RouteParameters<Next> : never)
```

Comment:
Intersection with `never` is `never`. It might not be possible to reach that condition, but better safe than sorry:

```suggestion
                (Rest extends `${GetRouteParameter<Rest>}${infer Next}` ? RouteParameters<Next> : unknown)
```

Reply:
done. though i usually use `never` in conditional types where the second condition shouldn't be reachable. is there a reason i should use `unknown` instead in cases like this?

Reply:
As I said above, you are intersecting a mapped type with the result of the conditional. Any intersection with `never` produces `never`, while any intersection with `unknown` removes `unknown`: 
```ts
type T = { x: number } & never; // never
type U = { x: number } & unknown; // { x: number }
```
This behavior is the inverse of a union, where any union with `unknown` produces `unknown`, while any union with `never` removes `never`:
```ts
type T = { x: number } | unknown; // unknown;
type U = { x: number } | never; // never;
```

Generally you *do* want the "unused" side of a conditional to be `never`, since conditional types are distributive:

```ts
type X<S extends string> = S extends "a" ? 0 : S extends "b" ? 1 : never;
type T = X<"a" | "b">; // 0 | 1 (since the conditional distributed over the union)
```

My concern was that if `Rest` was somehow the empty string, then you would end up in the false branch and you would end up with `{ [P in GetRouteParameter<Rest>]: string } & never` which would produce `never`. That could happen if the route string was `/:foo/:`, which *is* a valid route according to http://forbeslindesay.github.io/express-route-tester/:

![image](https://user-images.githubusercontent.com/3902892/122461575-5d12e580-cf68-11eb-8a9f-5b2ab0dc4669.png)

Switching from `never` to `unknown` *could* be problematic if the type for the route was a union like
`` "/:foo" | "/:" ``, since the distribution of `Rest` would result in `"foo" | unknown`, which degenerates to `unknown`. However, its unlikely that you'll encounter such a case since most routes are defined using a literal string.

If it is problematic, there are other ways to structure the type that don't rely on using an intersection type that could degenerate to `never`.

For example:

```ts
type SplitRoute<S extends string> =
    S extends `${infer L}/${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    S extends `${infer L}-${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    S extends `${infer L}.${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    [S];

type MapRoute<Segments extends string[]> = {
    [P in keyof Segments]: Segments[P] extends string ? ExtractRouteParameter<Segments[P]> : never;
}[number];

type ExtractRouteParameter<S extends string> = 
    S extends `${string}:${infer ParamName}?` ? [ParamName, true] :
    S extends `${string}:${infer ParamName}` ? [ParamName, false] :
    never;

type RouteParameters<Route extends string> =
    string extends Route ? ParamsDictionary : // degenerate case: RouteParameter<string>
    Route extends `${string}(${string}` ? ParamsDictionary : // regex unsupported
    [MapRoute<SplitRoute<Route>>] extends [infer Segments] ?
        { [P in Extract<Segments, [string, true]>[0]]?: string } &
        { [P in Extract<Segments, [string, false]>[0]]: string } :
    never;
```

[Typescript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACghgJzgWwM4BECWBjYmD2AdoiFALxQBKE2+CAJgDyrAKaEDmANFHISAD4A3AFgAUONCQoAZTAAbTMEr4ArsAiMZUCAA8Nheqigs2nAeXFRrsnfoiHjAAwAkAb3YAzCAigAZAF8AencvHyoApygAfigAbQA6JLlFZTUNRj8BHiSElKUVdU1KAQBdKAAuKxttPQMjKFcPQm9fQIBaUJbwykiY+Nz8tKLM7KhBhQL04rLK6uta+0dGrtb-AITVnr7YxOTJ4YysnP3UwoyS8qqxG3iZUtEJMSloAFk4MHPNGQgOZAdgMY6g4GqZ2Bw4qULBQ3PN4jAoOwoABrCAgfCeWS-f6EQGlCpYv4A1BxGDlYHLMGcfoAUX0SFwX3gSH+GgQWmxxNJUMqUEIEAAbj5HgE4oRVMgAEY+B7iSTgaB01hwRnTZkoCBsrR2erGKkcaFQOGLXUrNz6gIVLa+dXIAByGoC0Siu1tDv+PFYqggV2NOpBzncFqtzTWbsdLvhiBQ7ogPE8cHkqB9cxuNn5QoQj3l0iZ0dZPlQjC+-sprHB0Lh+tLDRLsVtGBweCIJFTtxLFIaTQtAAog+XODtYPnG7gCMQEKRrrc4u9PtMtAcvsXpgJZp3jHEwr4fkTcahytE4bc3PDEYQoEqGcAOXvATw4vrPQhvVC4gAGUqlaIE6sBKAAGTHjYp6kuel70iqN67ji97xE+UAJkmPoCB+X6-gOHBQP+07poKwpymIsJptYLxQAA+uQ8TAdY5HkQAKhALCMHRdKQLg9EKiuRS2pqhaMAAREEFSePg+DtBUkqINEAljKeon4BhZjcFAUkID+JiYdha5wrKYgBIRQRBFAADECC-OOUCMSwUAABYQPIkC+C8qDiEZUDIOwmBQGRGgsOCiEshAADudDIoF+DIHZwDAGAqAVEZHBKLZqiSgktDIEETGeLoWWoDlQSSvI+CSkEyBwCwPhBGAKrInAHBMUELztMg+D0A5QSoAg2BBHRfmAnRQTsO1ugJICJl+AAHAArAAzOImDIGAdDAFAp4AJKoHa+EIDwm0AIL8NhEVRQJAAC2W5c1rXtfIAmPGRdHWTe9E1sYp41ag8U+S+0ABIa9EPQqFHkWxNDAJxkCMPtuCqImPBg7gED0JWJHxJt22ZowiMaCj5JLA0cReimuwYzt0Ow4m64E5uxOHmtUCfd9xPHQSH0Vd9SHJgjujsbjBI48jPAw8AcPyASIti6zcJxGTWOS1T+OmkTv30+zX0ElzcYQXzyMC7z4NC1ACvi8blPyNLaOy6gh0gNjBtI3jb3xHT-TW7bFOi4rzsq6+-Tq8zv2swzTOa4m3M64b9D67r9DC+bEvm5bM4Hfwnti9Tyuu7EAdh8hPOxzHUfx17psm8nNhxILcdm6XSsBvEJsF1HauMxzBIszhcK54h4fa9XReOyXYuJ6X2GPO5JkguZSVEFZTGrfZjnhC5QA)

Reply:
You can even handle the basic case of matching parens (but not balanced or escaped parens, unfortunately) with this minor change to `SplitRoute`:
```ts
type SplitRoute<S extends string> =
    S extends `${infer L}(${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R extends `${string})${infer RR}` ? RR : "">] :
    S extends `${infer L}/${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    S extends `${infer L}-${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    S extends `${infer L}.${infer R}` ? [...SplitRoute<L>, ...SplitRoute<R>] :
    S extends "" ? [] : [S];
```